services/device-registry/src/main/java/org/eclipse/hono/deviceregistry/FileBasedTenantService.java

```diff
@@ -0,0 +1,392 @@
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+
+import org.eclipse.hono.service.tenant.BaseTenantService;
+import org.eclipse.hono.util.TenantResult;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.json.DecodeException;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+/**
+ * A tenant service that keeps all data in memory but is backed by a file.
+ * <p>
+ * On startup this adapter loads all registered tenants from a file. On shutdown all tenants kept in memory are written
+ * to the file.
+ */
+@Repository
+public final class FileBasedTenantService extends BaseTenantService<FileBasedTenantsConfigProperties> {
+
+    /**
+     * The name of the JSON property containing the tenant ID.
+     */
+    public static final String FIELD_TENANT = "tenant";
+    /**
+     * The name of the JSON property containing the tenant data.
+     */
+    public static final String FIELD_DATA = "data";
+
+    // <tenantId, tenantPropertyValueObject>
+    private Map<String, JsonObject> tenants = new HashMap<>();
+    // <tenantId, <adapterType, adapterDataObject>>
+    private Map<String, Map<String, JsonObject>> tenantsAdapterConfigurations = new HashMap<>();
+    private boolean running = false;
+    private boolean dirty = false;
+
+    @Autowired
+    @Override
+    public void setConfig(final FileBasedTenantsConfigProperties configuration) {
+        setSpecificConfig(configuration);
+    }
+
+    @Override
+    protected void doStart(Future<Void> startFuture) {
+
+        if (running) {
+            startFuture.complete();
+        } else {
+
+            if (!getConfig().isModificationEnabled()) {
+                log.info("modification of registered tenants has been disabled");
+            }
+
+            if (getConfig().getFilename() == null) {
+                log.debug("tenant manager filename is not set, no tenant information will be loaded");
+                running = true;
+                startFuture.complete();
+            } else {
+                checkFileExists(getConfig().isSaveToFile()).compose(ok -> {
+                    return loadTenantData();
+                }).compose(s -> {
+                    if (getConfig().isSaveToFile()) {
+                        log.info("saving tenants to file every 3 seconds");
+                        vertx.setPeriodic(3000, tid -> {
+                            saveToFile();
+                        });
+                    } else {
+                        log.info("persistence is disabled, will not save tenants to file");
+                    }
+                    running = true;
+                    startFuture.complete();
+                }, startFuture);
+            }
```

Comment:
we are doing all this file handling logic over and over again, shouldn't we start to pull this up into a super class or refactor it into a separate helper class?

Reply:
Yes, definitely. Will do that separately from this PR, though, since this not only affects the tenant API implementation.